src/topology_local.ts

```diff
@@ -173,32 +172,33 @@ export class TopologyLocal {
             throw new Error(this.logging_prefix + "Topology is already running.");
         }
         log.logger().log(this.logging_prefix + "Local topology started");
+        // spouts pass internal exceptions to errorCallback
         for (let spout of this.spouts) {
             spout.run();
         }
         this.isRunning = true;
     }
 
-    /** Sends pause signal to all spouts */
+    /** Sends pause signal to all spouts. Each spout.pause is idempotent  */
     pause(callback: intf.SimpleCallback) {
-        try {
-            if (!this.isInitialized) {
-                throw new Error(this.logging_prefix + "Topology not initialized and cannot be paused.");
-            }
-            for (let spout of this.spouts) {
-                spout.pause();
-            }
-            this.isRunning = false;
-        } catch (e) {
-            callback(e);
+        if (!this.isInitialized) { 
+            return callback(new Error(this.logging_prefix + "Topology not initialized and cannot be paused."));
         }
-        callback();
+        // spouts pass internal exceptions to errorCallback
+        for (let spout of this.spouts) {
+            spout.pause();
```

Comment:
what if `spout.pause()` raises an exception? you removed try/catch

Reply:
Because everything in spout.pause is in try - catch

Reply:
I don't see that :)
https://github.com/rupnikj/qtopology/blob/f7342b845a9b4c077339c1f370c1f69488a7c9bb/src/topology_local.ts#L183

what am I missing?

Reply:
https://github.com/rupnikj/qtopology/blob/master/src/topology_local_inprocess.ts#L272
each spout.pause cannot throw